### PR TITLE
Refactor some ta mutex usage

### PR DESCRIPTION
**Pull request by jenswi-linaro**

### CODE REVIEW

File: core/arch/arm/include/kernel/thread.h [hunk 1]
```diff
@@ -247,6 +247,7 @@ struct thread_specific_data {
 #ifdef CFG_CORE_DEBUG_CHECK_STACKS
 	bool stackcheck_recursion;
 #endif
+	unsigned int recursion_depth;
```
Comment (jforissier) [inline core/arch/arm/include/kernel/thread.h:250]: `syscall_recursion` to better reflect the `CFG_` name?

**Review by jforissier — COMMENTED** [review]
For commit "core: add CFG_CORE_MAX_SYSCALL_RECURSION":
One suggestion below, anyways:
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

**Review by jforissier — COMMENTED** [review]
Reviewing "core: refactor locking during user TA load".

I fail to see which sequence of events can lead to a deadlock, can you give an example?

File: core/arch/arm/include/kernel/thread.h (hunk 1 above)
Comment (jenswi-linaro) [reply core/arch/arm/include/kernel/thread.h:250]: Makes sense, I'll fix.

File: core/kernel/tee_ta_manager.c [hunk 2]
```diff
@@ -623,29 +634,20 @@ static TEE_Result tee_ta_init_session(TEE_ErrorOrigin *err,
...
-	 * down into smaller pieces.
-	 */
-
-
 	mutex_lock(&tee_ta_mutex);
 	s->id = new_session_id(open_sessions);
 	if (!s->id) {
 		res = TEE_ERROR_OVERFLOW;
-		goto out;
+		goto err_free_s;
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/kernel/tee_ta_manager.c:641]: Thanks, I'll fix.

**Comment by jenswi-linaro** [comment]
> Reviewing "core: refactor locking during user TA load".
> 
> I fail to see which sequence of events can lead to a deadlock, can you give an example?

If `pgt_alloc()` cannot provide all needed page tables it will sleep until some have been freed and try again. If this happens when `tee_ta_mutex` is locked by the same thread it goes rather quiet quickly. One way of demonstrating this is to set `PGT_CACHE_SIZE` to `3` and run `xtest 1013`.

**Comment by jforissier** [comment]
Now I get it. The debug traces are not easy to understand because `__wq_rpc sleep` always prints the mutex address so it is not clear if we sleep trying to lock the mutex or waiting on a cond var which uses this mutex.

**Review by jforissier — APPROVED** [review]
LGTM.
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

**Comment by jenswi-linaro** [comment]
> Now I get it. The debug traces are not easy to understand because `__wq_rpc sleep` prints only the mutex address so it is not clear if we sleep trying to lock the mutex or waiting on a cond var which uses this mutex.

`CFG_MUTEX_DEBUG=y` can help a bit here.

**Pull request merged**